### ISSUE TITLE
Praise and questions and then some :)

### ISSUE THREAD

**Issue by Christian-Schiffer** [hherzl/CatFactory#4]
Only had time for a short read through yet but from what I can see, I can scaffold everything from the data access layer and repositories via business objects to the web.api controllers based on my data models, even typescript? wow! I currently use TypewriterX to scaffold models and services for angular from my data models and API controllers, respectively, that amazing success has given me the urge to go even further and that is how I ended up here.

From the short read through it seams its easy to learn your templating DSL and a quick look at your GitHub repository tells me you are planning full angular support as well. I do not have any questions yet, I am sure I will when I get to it, but I do wish to add some praise and thanks for this scaffolding framework. great job!!! Thumbs Up | :thumbsup: Thumbs Up | :thumbsup: Thumbs Up | :thumbsup: 

Ops, there came the first question to mind, is there any get started video to be found yet? Cool | :cool: 

Also, I checked a bit further, does it work already? 

PS: I by the way tried to send you a pull request since I added public GetKebabCase and private HasUpper and HasLower methods to the CatFactory.NamingConvention class as well as unit tests to check if they work as well as fixed the GetSnakeCase method so it would also split on uppercase and added additional unit tests for it. I think I failed though, never used git as source control before, only TFS and Azure DevOps... Please advice...

Cheers!

**Comment by hherzl** [comment]
Hi.

Clone this repository, make changes then commit and push.

Do you have Git tools for Visual Studio?

I'll take a look on Kebab Kase.

Let me know if this reply is useful?

Regards

**Comment by Christian-Schiffer** [comment]
Thanks, yeah great, I'll try that... I thought I did though.. Maybe its because I use VS 2019.2.2?... I tried again now, http status 403, forbidden, this is the error message I got:

Opening repositories:
C:\Users\David\Source\Repos\CatFactory
Commit 64b47676 created locally in repository C:\Users\David\Source\Repos\CatFactory
Pushing master
Remote: Permission to hherzl/CatFactory.git denied to Christian-Schiffer.
Error encountered while pushing to the remote repository: Git failed with a fatal error.
unable to access 'https://github.com/hherzl/CatFactory/': The requested URL returned error: 403
Pushing to https://github.com/hherzl/CatFactory

Maybe I should use VS code instead?

**Comment by Christian-Schiffer** [comment]
Anyway this is the code...

`using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace CatFactory
{
    /// <summary>
    /// Provides helper methods for naming conventions
    /// </summary>
    public static class NamingConvention
    {
        // ReSharper disable once UnusedMember.Local
        private static bool IsLower(string source)
            => source.ToLower() == source;

        private static bool IsUpper(string source)
            => source.ToUpper() == source;

        private static bool HasUpper(string source)
            => source.Any(char.IsUpper);

        private static bool HasLower(string source)
            => source.Any(char.IsLower);

        /// <summary>
        /// Gets a camelCase style string
        /// </summary>
        /// <param name="source">Source string</param>
        /// <returns>A <see cref="string"/> that represents camelString for source string</returns>
        public static string GetCamelCase(string source)
        {
            if (source.Length == 0)
                return string.Empty;

            if (source.Contains(" "))
            {
                var pieces = source.Split(' ');

                var name = new StringBuilder();

                for (var i = 0; i < pieces.Length; i++)
                {
                    var item = pieces[i];

                    if (item.Length == 0)
                        continue;

                    name.Append(i == 0 ? item[0].ToString().ToLower() : item[0].ToString().ToUpper());
                    name.Append(item.Substring(1));
                }

                return name.ToString();
            }
            else if (source.Contains("_"))
            {
                var pieces = source.Split('_');

                var name = new StringBuilder();

                for (var i = 0; i < pieces.Length; i++)
                {
                    var item = pieces[i];

                    if (item.Length == 0)
                        continue;

                    name.Append(i == 0 ? item[0].ToString().ToLower() : item[0].ToString().ToUpper());
                    name.Append(item.Substring(1));
                }

                return name.ToString();
            }
            else
            {
                if (source.Length == 1 || IsUpper(source))
                    return source.ToLower();

                return string.Format("{0}{1}", source[0].ToString().ToLower(), source.Substring(1)).Replace("_", string.Empty).Replace(".", string.Empty);
            }
        }

        /// <summary>
        /// Gets a PamelCase style string
        /// </summary>
        /// <param name="source">Source string</param>
        /// <returns>A <see cref="string"/> that represents PascalString for source string</returns>
        public static string GetPascalCase(string source)
        {
            if (source.Length == 0)
                return string.Empty;

            source = source.Replace("  ", " ").Trim();

            var name = new StringBuilder();

            if (source.Contains("_"))
            {
                var pieces = source.Split('_');

                foreach (var item in pieces)
                {
                    if (item.Length == 0)
                        continue;

                    name.Append(item[0].ToString().ToUpper());
                    name.Append(IsUpper(item) ? item.Substring(1).ToLower() : item.Substring(1));
                }
            }
            else if (source.Contains("."))
            {
                var pieces = source.Split('.');

                foreach (var item in pieces)
                {
                    if (item.Length == 0)
                        continue;

                    name.Append(item[0].ToString().ToUpper());
                    name.Append(IsUpper(item) ? item.Substring(1).ToLower() : item.Substring(1));
                }
            }
            else if (source.Contains(" "))
            {
                var pieces = source.Split(' ');

                foreach (var item in pieces)
                {
                    if (item.Length == 0)
                        continue;

                    name.Append(item[0].ToString().ToUpper());
                    name.Append(item.Substring(1).ToLower());
                }
            }
            else
            {
                if (source.Length == 1)
                {
                    name.Append(source.ToUpper());
                }
                else
                {
                    name.Append(source[0].ToString().ToUpper());
                    name.Append(IsUpper(source) ? source.Substring(1).ToLower() : source.Substring(1));
                }
            }

            return name.ToString();
        }

        /// <summary>
        /// Gets a Snake_Case style string
        /// </summary>
        /// <param name="source">Source string</param>
        /// <returns>A <see cref="string"/> that represents Snake_Case for source string</returns>
        public static string GetSnakeCase(string source)
        {
            if (source.Length == 0)
                return string.Empty;

            if (source.Contains("_"))
                return source;

            source = source.Replace("  ", " ").Trim();

            var name = new List<string>();

            if (source.Contains("."))
            {
                var pieces = source.Split('.');

                name.AddRange(pieces.Where(item => item.Length != 0));
            }
            else if (source.Contains(" "))
            {
                var pieces = source.Split(' ');

                name.AddRange(pieces.Where(item => item.Length != 0));
            }
            else if (HasUpper(source) & !IsUpper(source))
            {
                name.AddRange(Regex.Split(source, @"(?<!^)(?=[A-Z])"));
            }
            else
            {
                name.Add(source);
            }

            return string.Join("_", name);
        }

        /// <summary>
        /// Creates  a kebab-case style string
        /// </summary>
        /// <param name="source">Source string</param>
        /// <returns>A <see cref="string"/> that represents kebab-case for source string</returns>
        public static string GetKebabCase(string source)
        {
            if (source.Length == 0)
                return string.Empty;

            if (source.Contains("-"))
                return source;

            source = source.Replace("  ", " ").Trim();

            var name = new List<string>();

            if (source.Contains("."))
            {
                var pieces = source.Split('.');

                name.AddRange(pieces.Where(item => item.Length != 0));
            }
            else if (source.Contains(" "))
            {
                var pieces = source.Split(' ');

                name.AddRange(pieces.Where(item => item.Length != 0));
            }
            else if (source.Contains("_"))
            {
                var pieces = source.Split('_');

                name.AddRange(pieces.Where(item => item.Length != 0));
            }
            else if (HasUpper(source) & !IsUpper(source))
            {
                name.AddRange(Regex.Split(source, @"(?<!^)(?=[A-Z])"));
            }
            else
            {
                name.Add(source);
            }

            var retValue = string.Join("-", name).ToLower();
            return retValue;
        }

    }
}
`

And the tests:

`using Xunit;

namespace CatFactory.Tests
{
    public class NamingConventionTests
    {
        [Fact]
        public void TestCamelCase()
        {
            Assert.True(NamingConvention.GetCamelCase("FOO") == "foo");
            Assert.True(NamingConvention.GetCamelCase("Bar") == "bar");
            Assert.True(NamingConvention.GetCamelCase("zaz") == "zaz");
            Assert.True(NamingConvention.GetCamelCase("FooBarZaz") == "fooBarZaz");
            Assert.True(NamingConvention.GetCamelCase("foo bar zaz") == "fooBarZaz");
            Assert.True(NamingConvention.GetCamelCase("foo_bar_zaz") == "fooBarZaz");
        }

        [Fact]
        public void TestPascalCase()
        {
            Assert.True(NamingConvention.GetPascalCase("FOO") == "Foo");
            Assert.True(NamingConvention.GetPascalCase("Bar") == "Bar");
            Assert.True(NamingConvention.GetPascalCase("zaz") == "Zaz");
            Assert.True(NamingConvention.GetPascalCase("fooBarZaz") == "FooBarZaz");
            Assert.True(NamingConvention.GetPascalCase("foo.bar.zaz") == "FooBarZaz");
            Assert.True(NamingConvention.GetPascalCase("foo_bar_zaz") == "FooBarZaz");
            Assert.True(NamingConvention.GetPascalCase("foo bar zaz") == "FooBarZaz");
        }

        [Fact]
        public void TestSnakeCase()
        {
            Assert.True(NamingConvention.GetSnakeCase("schema_id") == "schema_id");
            Assert.True(NamingConvention.GetSnakeCase("schema.id") == "schema_id");
            Assert.True(NamingConvention.GetSnakeCase("schema id") == "schema_id");
            Assert.True(NamingConvention.GetSnakeCase("FooBarZaz") == "Foo_Bar_Zaz");
            Assert.True(NamingConvention.GetSnakeCase("foo bar zaz") == "foo_bar_zaz");
            Assert.True(NamingConvention.GetSnakeCase("foo_bar_zaz") == "foo_bar_zaz");
        }

        /// <summary>   Tests kebab case. </summary>
        ///
        /// <remarks>   Christian Schiffer, 12.02.2019. </remarks>
        [Fact]
        public void TestKebabCase()
        {
            Assert.True(NamingConvention.GetKebabCase("FOO") == "foo");
            Assert.True(NamingConvention.GetKebabCase("Bar") == "bar");
            Assert.True(NamingConvention.GetKebabCase("zaz") == "zaz");
            Assert.True(NamingConvention.GetKebabCase("FooBarZaz") == "foo-bar-zaz");
            Assert.True(NamingConvention.GetKebabCase("foo bar zaz") == "foo-bar-zaz");
            Assert.True(NamingConvention.GetKebabCase("foo-bar-zaz") == "foo-bar-zaz");
            Assert.True(NamingConvention.GetKebabCase("foo_bar_zaz") == "foo-bar-zaz");
        }
    }
}
`

**Comment by hherzl** [comment]
Hi.

I have added to you as collaborator for this repository.

Try to commit your changes, if you have any trouble please let me know.

Regards

**Comment by Christian-Schiffer** [comment]
Great thank you, I love your project and am happy to be able to contribute a little. I was able to commit the changes I created a few days ago today and also created a new change adding xml comments to follow up on the (new?) excellent commenting strategy. The terminology of github is also starting to make sense to me. (coming from TFS / Azure DevOps this was my first and second commit to github, ever) Have been looking at your projects here on github and on codeproject, great stuff and very much in line with what I am working with too.
